pkg/util/ioutil/ioutil.go

```diff
@@ -7,70 +7,63 @@ package ioutil
 import (
 	"context"
 	"io"
+	"sync/atomic"
 	"time"
 )
 
 // timeoutReader monitors the progress of the io.Reader Read method
-// and calls the cancel function when reading does not progress for
+// and calls the cancelFn function when reading does not progress for
 // the specified time.
 type timeoutReader struct {
-	t *time.Timer
-	r io.Reader
-	n chan int
+	total  atomic.Uint64 // Total number of bytes read until timeout.
+	reader io.Reader
+
+	timer   *time.Timer
+	timeout time.Duration
+
+	done   atomic.Bool
+	cancel context.CancelFunc
 }
 
 // Read implements the io.Reader Read interface.
 func (tr *timeoutReader) Read(p []byte) (int, error) {
-	n, err := tr.r.Read(p)
-
-	v := n
-	if err != nil && tr.t.Stop() {
-		select {
-		case <-tr.t.C:
-		default:
+	n, err := tr.reader.Read(p)
+	if !tr.done.Load() && tr.timer.Reset(tr.timeout) {
+		tr.total.Add(uint64(n))
+		if err != nil {
+			tr.cancel()
+			tr.done.Store(true)
 		}
-
-		// Negative value signals that error has
-		// occurred and the goroutine should terminate.
-		v = -1
 	}
-
-	select {
-	case tr.n <- v:
-	default:
-	}
-
 	return n, err
 }
 
 // TimeoutReader creates a new timeoutReader instance and starts
 // a goroutine that monitors the progress of reading from the given reader.
 // If no progress is made for the duration of the given timeout, then the
-// reader executes the given cancel function and terminates the underlying
+// reader executes the given cancelFn function and terminates the underlying
 // goroutine. This goroutine will also terminate when the given context
 // is canceled or if an error is returned by the given reader.
-func TimeoutReader(ctx context.Context, r io.Reader, timeout time.Duration, cancel func(uint64)) io.Reader {
-	tr := &timeoutReader{t: time.NewTimer(timeout), r: r, n: make(chan int)}
+func TimeoutReader(ctx context.Context, reader io.Reader, timeout time.Duration, cancelFn func(uint64)) io.Reader {
+	ctx, cancel := context.WithCancel(ctx)
+	tr := &timeoutReader{
+		reader:  reader,
+		timer:   time.NewTimer(timeout),
+		timeout: timeout,
+		cancel:  cancel,
+	}
 
 	go func() {
-		var total uint64 = 0
+		defer func() {
+			tr.done.Store(true)
+		}()
 
-		for {
-			select {
-			case n := <-tr.n:
-				switch {
-				case n < 0:
-					return
-				case n >= 0:
-					tr.t.Reset(timeout)
-					total += uint64(n)
-				}
-			case <-tr.t.C:
-				cancel(total)
-				return
-			case <-ctx.Done():
-				return
-			}
+		select {
+		case <-ctx.Done():
+			return
+		case <-tr.timer.C:
+			cancelFn(tr.total.Load())
```

Comment:
`cancelFn` might take some time to execute (hypothetically speaking), this could cause following error:

1. timer is fired
2. `cancelFn` is being now executed (this is goroutine A)
3.  reader continues to read data (this is goroutine B; assuming B is faster then A), it resets timer because there are more data to read
=> timer will continue ticking once again

This might not be big problem, but if would make sense to avoid it if possible. Could stopping/cleanup be done before calling cancel?

Reply:
Point 3 is invalid because the `select` statement is not in the loop. In other words, if the timer is fired and the value is read, then reset, there will be no more goroutine waiting to read from it.

Reply:
> then reset, there will be no more goroutine waiting to read from it.

Agree. What will happen, in this case, is that timer goroutine will keep alive for duration of timeout (because timer got restarted), but nobody will ever going to read from it (goroutine has returned). In other words, some resources associated to this timer will be unnecessarily kept.